File: android/src/main/java/com/valentingrigorean/arcgis_maps_flutter/loadable/LoadableNativeHandler.java
package com.valentingrigorean.arcgis_maps_flutter.loadable;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import com.esri.arcgisruntime.loadable.LoadStatusChangedEvent;
import com.esri.arcgisruntime.loadable.LoadStatusChangedListener;
import com.esri.arcgisruntime.loadable.Loadable;
import com.valentingrigorean.arcgis_maps_flutter.Convert;
import com.valentingrigorean.arcgis_maps_flutter.flutterobject.BaseNativeHandler;

import io.flutter.plugin.common.MethodChannel;

public class LoadableNativeHandler extends BaseNativeHandler<Loadable> implements LoadStatusChangedListener {

    public LoadableNativeHandler(Loadable loadable) {
        super(loadable);
        loadable.addLoadStatusChangedListener(this);
    }

    @Override
    protected void disposeInternal() {
        super.disposeInternal();
        getNativeHandler().removeLoadStatusChangedListener(this);
    }

    @Override
    public boolean onMethodCall(@NonNull String method, @Nullable Object args, @NonNull MethodChannel.Result result) {
        switch (method) {
            case "loadable#getLoadStatus":
                result.success(getNativeHandler().getLoadStatus().ordinal());
                return true;
            case "loadable#getLoadError":
                result.success(Convert.arcGISRuntimeExceptionToJson(getNativeHandler().getLoadError()));
                return true;
            case "loadable#cancelLoad":
                getNativeHandler().cancelLoad();
                result.success(null);
                return true;
            case "loadable#loadAsync":
                getNativeHandler().loadAsync();
                getNativeHandler().addDoneLoadingListener(new DoneListener(result));
                return true;
            case "loadable#retryLoadAsync":
                getNativeHandler().retryLoadAsync();
                getNativeHandler().addDoneLoadingListener(new DoneListener(result));
                return true;
        }
        return false;
    }

    @Override
    public void loadStatusChanged(LoadStatusChangedEvent loadStatusChangedEvent) {
        sendMessage("loadable#loadStatusChanged", loadStatusChangedEvent.getNewLoadStatus().ordinal());
    }

    private class DoneListener implements Runnable {
        private final MethodChannel.Result result;

        private DoneListener(MethodChannel.Result result) {
            this.result = result;
        }

        @Override
        public void run() {
            getNativeHandler().removeDoneLoadingListener(this);
            if (!isDisposed()) {
                result.success(null);
            }
        }
    }
}

File: android/src/main/kotlin/com/valentingrigorean/arcgis_maps_flutter/loadable/LoadableNativeHandler.kt
package com.valentingrigorean.arcgis_maps_flutter.loadable

import com.arcgismaps.LoadStatus
import com.arcgismaps.Loadable
import com.valentingrigorean.arcgis_maps_flutter.convert.toFlutterJson
import com.valentingrigorean.arcgis_maps_flutter.convert.toFlutterValue
import com.valentingrigorean.arcgis_maps_flutter.flutterobject.BaseNativeHandler
import io.flutter.plugin.common.MethodChannel
import kotlinx.coroutines.flow.launchIn
import kotlinx.coroutines.flow.onEach
import kotlinx.coroutines.launch

class LoadableNativeHandler(loadable: Loadable) : BaseNativeHandler<Loadable>(loadable) {
    init {
        loadable.loadStatus.onEach { status ->
            sendMessage("loadable#loadStatusChanged", status.toFlutterValue())
        }.launchIn(scope)
    }

    override fun onMethodCall(method: String, args: Any?, result: MethodChannel.Result): Boolean {
        when (method) {
            "loadable#getLoadStatus" -> {
                result.success(nativeHandler.loadStatus.value.toFlutterValue())
                return true
            }

            "loadable#getLoadError" -> {
                val loadError = nativeHandler.loadStatus.value as? LoadStatus.FailedToLoad
                if (loadError != null) {
                    result.success(loadError.error.toFlutterJson())
                } else {
                    result.success(null)
                }
                return true
            }

            "loadable#cancelLoad" -> {
                nativeHandler.cancelLoad()
                result.success(null)
                return true
            }

            "loadable#loadAsync" -> {
                scope.launch {
                    nativeHandler.load().onSuccess { result.success(null) }
                        .onFailure { result.error("loadable#loadAsync", it.message, it) }
                }
                return true
            }

            "loadable#retryLoadAsync" -> {
                scope.launch {
                    nativeHandler.retryLoad().onSuccess { result.success(null) }
                        .onFailure { result.error("loadable#retryLoadAsync", it.message, it) }
                }
                return true
            }
        }
        return false
    }
}